File: pennyfarthing_scripts/prime/version_sentinel.py
"""Version sentinel detection for auto-update in prime.

MSSCI-14698: Reads .pennyfarthing/.installed-version sentinel and compares
against current package version. On mismatch, signals that auto-update
is needed.

STUB: Implementation needed by Dev (story 98-1).
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path


SENTINEL_FILENAME = ".installed-version"


@dataclass
class VersionCheckResult:
    """Result of version mismatch check."""

    needs_update: bool
    installed_version: str | None
    package_version: str


def read_sentinel_version(project_root: Path) -> str | None:
    """Read version from .pennyfarthing/.installed-version sentinel.

    Args:
        project_root: Project root directory

    Returns:
        Version string, or None if sentinel doesn't exist or is empty
    """
    # STUB: Not yet implemented — tests should fail on assertions
    raise NotImplementedError("read_sentinel_version not implemented")


def get_package_version(project_root: Path) -> str:
    """Get the current package version from VERSION file or package.json.

    Args:
        project_root: Project root directory

    Returns:
        Version string
    """
    # STUB: Not yet implemented
    raise NotImplementedError("get_package_version not implemented")


def check_version_mismatch(
    project_root: Path,
    *,
    package_version: str | None = None,
) -> VersionCheckResult:
    """Compare sentinel version against package version.

    Args:
        project_root: Project root directory
        package_version: Override package version (for testing).
            If None, reads from VERSION file.

    Returns:
        VersionCheckResult with needs_update flag
    """
    # STUB: Not yet implemented — tests should fail on assertions
    raise NotImplementedError("check_version_mismatch not implemented")

File: packages/core/src/cli/utils/version-sentinel.ts
/**
 * Version sentinel file utilities for MSSCI-14698.
 *
 * Manages the .pennyfarthing/.installed-version sentinel file that tracks
 * which version of Pennyfarthing is installed. Used by prime to detect
 * version mismatches and trigger auto-updates.
 *
 * STUB: Implementation needed by Dev (story 98-1).
 */

/**
 * Sentinel filename within .pennyfarthing/
 */
export const SENTINEL_FILENAME = '.installed-version';

/**
 * Write the version sentinel file to .pennyfarthing/.installed-version.
 *
 * Called by init.ts after successful initialization and by update.ts
 * after successful update.
 *
 * @param projectRoot - Project root directory
 * @param version - Version string to write
 * @param options - Options (dryRun skips write)
 */
export function writeVersionSentinel(
  _projectRoot: string,
  _version: string,
  _options?: { dryRun?: boolean }
): void {
  // STUB: Not yet implemented — tests should fail on assertions
  throw new Error('writeVersionSentinel not implemented');
}

/**
 * Read the version from the sentinel file.
 *
 * @param projectRoot - Project root directory
 * @returns Version string, or null if sentinel doesn't exist or is empty
 */
export function readVersionSentinel(_projectRoot: string): string | null {
  // STUB: Not yet implemented — tests should fail on assertions
  throw new Error('readVersionSentinel not implemented');
}
